radar/src/radar.rs

```diff
@@ -0,0 +1,603 @@
+            &frame.entity_id,
+            &position,
+            &radar_receiver,
+            &old_contacts,
+            &all_positions,
+        );
+
+        for update in updates {
+            match update {
+                RadarContactDelta::Add(rc) => {
+                    let new_subject = ResProtocolRequest::New(format!(
+                        "{}.{}",
+                        resource_id.to_string(),
+                        RADAR_CONTACTS
+                    ))
+                    .to_string();
+                    let payload = serde_json::json!({ "params": rc });
+                    publish_message(ctx, &new_subject, payload)?;
+                }
+                RadarContactDelta::Remove(rid) => {
+                    let delete_subject = ResProtocolRequest::Delete(format!(
+                        "{}.{}",
+                        resource_id.to_string(),
+                        RADAR_CONTACTS
+                    ))
+                    .to_string();
+                    let payload = serde_json::json!({ "params": {"rid": rid.replace(":", ".")}});
+                    publish_message(ctx, &delete_subject, payload)?;
+                }
+                RadarContactDelta::Change(rid, rc) => {
+                    let set_subject = &format!("call.{}.set", rid);
+                    let payload = serde_json::json!({ "params": rc });
+                    publish_message(ctx, set_subject, payload)?;
+                }
+            }
+        }
+    }
+    Ok(vec![])
+}
+
+/// Helper function used to publish a payload on a specified subjct
+fn publish_message(
+    ctx: &CapabilitiesContext,
+    subject: &str,
+    payload: serde_json::value::Value,
+) -> CallResult {
+    if ctx
+        .msg()
+        .publish(subject, None, &serde_json::to_vec(&payload)?)
+        .is_err()
+    {
+        Err("Error publishing message".into())
+    } else {
+        Ok(vec![])
+    }
+}
+
+/// Function to compute all changes to a contact list needed given a resources id, current position,
+/// radar receiver, all old contacts, and a map of all entity positions that are published.
+/// Changes are in the form of RadarContactDeltas, either specifying to Add, Remove, or Change a contact.
+fn radar_updates(
+    entity_id: &str,
+    current_position: &Position,
+    radar_receiver: &RadarReceiver,
+    old_contacts: &HashMap<String, RadarContact>,
+    all_positions: &HashMap<String, Position>,
+) -> Vec<RadarContactDelta> {
+    let mut deltas: Vec<RadarContactDelta> = Vec::new();
+    let contacts: Vec<String> = old_contacts
+        .values()
+        .map(|rc| rc.clone().entity_id)
+        .collect();
+    for (k, v) in all_positions.iter() {
+        if contacts.contains(k) {
+            let mut rid: String = "".to_string();
+            for (key, val) in old_contacts.iter() {
+                if val.entity_id == *k {
+                    rid = key.to_string();
+                    break;
+                }
+            }
```

Comment:
I'm not very happy about my strategy here, technically it works but I think there's a more efficient way to accomplish this than getting a vector of the values in old_contacts and reverse engineering it.